docs: Add docker approach guide for dev setup

lukasaric opened the pull request:
Closes #1008

## Motivation and Resolution

The primary motivation for this PR is to improve the overall contribution guide. Furthermore, this PR introduces the `docker` approach so the contributor can easily set up the local environment.

### RPC version (if applicable)

_n.a_

## Usage related changes

When cloning the repo for the first time, you'll have a detailed guide on how to set docker and run the desired command from the tag exposed via this [link](https://hub.docker.com/r/shardlabs/starknet-devnet-rs/tags)

## Development related changes

- Clone the repo
- run `npm run install`
- Open the docker app on your machine (or run it via terminal if it's installed via brew or some other package manager)
- Run: ```docker run -p 5050:5050 shardlabs/starknet-devnet-rs:latest```
- Run `npm run test`

## Screenshot:
![Uploading Screenshot 2024-03-14 at 10.18.11.png…]()



## Checklist:

- [x] Performed a self-review of the code
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Linked the issues which this PR resolves
- [x] Documented the changes in code (API docs will be generated automatically)
- [x] Updated the tests
- [x] All tests are passing

lukasaric commented:
> Thanks! I think it would be useful to add a message indicating a devnet is not started when trying to run tests? Instead of this error message: `Error: Jest: Got error running globalSetup - /Users/tabaktoni/Work/starknet.js/__tests__/config/jestGlobalSetup.ts, reason: Setting Account using all known strategies failed, provide basic test parameters`

Hey!
Well, above the error this is shown:
```
Determining test suites to run...Global Test Setup Started
 TEST_ACCOUNT_ADDRESS env is not provided!
TEST_ACCOUNT_PRIVATE_KEY env is not provided!
Basic test parameters are missing, Auto Setup Started
```
Should this be sufficient? It indicates that you did not set things up properly

tabaktoni commented:
The idea is that one does not need to set up env.
If you have local devnet running it will auto setup if not it will fail with an error.
The solution is to modify jestGlobalSetup so that it prints what to do to setup local devnet if auto setup failed

lukasaric commented:
> The idea is that one does not need to set up env. If you have local devnet running it will auto setup if not it will fail with an error. The solution is to modify jestGlobalSetup so that it prints what to do to setup local devnet if auto setup failed

I extracted the devnet detector as a separate file and added a screenshot via description to see the detailed message that the local devnet is not running.
Moreover, I can extract other stuff (this can be a separate PR/issue) within the global jest setup so that future devs/contributors have isolated files. It will be easier to understand what's going on.

ivpavici commented on __tests__/config/jestGlobalSetup.ts:
```diff
@@ -216,13 +157,9 @@ const executeStrategy = async () => {
-    } else {
-      console.log('Detected Devnet-PY');
-    }
+    console.log(devnetStrategy.isRS ? 'Detected Devnet-RS' : 'Detected Devnet-PY');
```
@penovicp I think we can remove the Devnet-PY stuff?

lukasaric replied:
If we don't need it, I'll just log the rs devnet detection indication within the execute directly

lukasaric commented on __tests__/config/helpers/localDevnetDetector.ts:
```diff
@@ -0,0 +1,72 @@
+    return this.strategy;
+  }
+
+  private async isLocalDevnet(): Promise<boolean> {
```
Do we need this check at all? Can we just do the `isRsDevnet` check and if that one fails, just throw the error that your local RS devnet is not running with the following instructions?

tabaktoni replied:
We need to remove python devnet and sequencer, but my initial plan was to do it in next pr.
Anyway we can do it also in this one.
Yes isLocalDevnet is redundant here

lukasaric replied:
I'll remove the `isLocalDevnet`. As for the py devnet and sequencer, we can do that in a separate MR...you can maybe create an issue for the sequencer removal for that one ;)

lukasaric commented on CONTRIBUTING.md:
````diff
@@ -34,7 +34,8 @@ Or run tests in watch mode:
 ```
 
-By default the tests are executed in your local Devnet. If you want to use a specific rpc node, you have to set some global variables before executing the tests :
+By default the tests are executed in your local Devnet. If you want to use a specific
````
@tabaktoni @ivpavici We can also remove this completely and just rely on the docker approach. By doing that, we do not confuse the contributor with other options that can easily break

tabaktoni reviewed (approved):
LGTM!
Let's split it into two, I would like to clear the PR mess for now to make it more manageable

penovicp commented on CONTRIBUTING.md:
```diff
@@ -43,6 +44,16 @@ export TEST_ACCOUNT_ADDRESS=0x065A822f0000000000000000000000000c26641
+The recommended and more straightforward approach is to go with the docker.
+You just need to do the following steps:
+
+- Install the [docker](https://docs.docker.com/desktop/install/mac-install/) (You can install it via `brew` or some other package manager)
+- Run the `docker` on your machine (Open the application).
```
```suggestion
- Install [Docker](https://docs.docker.com/engine/install/) (it can also be installed via a package manager, e.g. `brew` for Mac)
- Run `Docker` on your machine (open the application).
```

penovicp commented on __tests__/config/helpers/localDevnetDetector.ts:
```diff
@@ -0,0 +1,73 @@
+Local devnet is not running. In order to properly run it you need to do the following: \n
+  - Go to the: https://hub.docker.com/r/shardlabs/starknet-devnet-rs/tags
+  - Find the latest tag and copy the "docker pull" command
+  - Run the docker on your machine
+  - Run the: "docker pull shardlabs/starknet-devnet-rs:latest"
```
```suggestion
  - Run Docker on your machine
  - Run the command: "docker pull shardlabs/starknet-devnet-rs:latest"
```

penovicp reviewed (approved):
lgtm

Left two nitpicky suggestions where using `Docker` as a proper noun feels more natural.
